java/ql/src/Telemetry/AutomodelApplicationModeCharacteristics.qll

```diff
@@ -0,0 +1,474 @@
+  UnexploitableExistsCharacteristic() { this = "unexploitable (existence-checking boolean method)" }
+
+  override predicate appliesToEndpoint(Endpoint e) {
+    not ApplicationCandidatesImpl::isSink(e, _) and
+    exists(Callable callable |
+      callable = ApplicationCandidatesImpl::getCallable(e) and
+      callable.getName().toLowerCase() = ["exists", "notexists"] and
+      callable.getReturnType() instanceof BooleanType
+    )
+  }
+}
+
+/**
+ * A negative characteristic that indicates that an endpoint is an argument to an exception, which is not a sink.
+ */
+private class ExceptionCharacteristic extends CharacteristicsImpl::NotASinkCharacteristic {
+  ExceptionCharacteristic() { this = "exception" }
+
+  override predicate appliesToEndpoint(Endpoint e) {
+    ApplicationCandidatesImpl::getCallable(e).getDeclaringType().getASupertype*() instanceof
+      TypeThrowable
+  }
+}
+
+/**
+ * A negative characteristic that indicates that an endpoint is a MaD taint step. MaD modeled taint steps are global,
+ * so they are not sinks for any query. Non-MaD taint steps might be specific to a particular query, so we don't
+ * filter those out.
+ */
+private class IsMaDTaintStepCharacteristic extends CharacteristicsImpl::NotASinkCharacteristic {
+  IsMaDTaintStepCharacteristic() { this = "taint step" }
+
+  override predicate appliesToEndpoint(Endpoint e) {
+    FlowSummaryImpl::Private::Steps::summaryThroughStepValue(e, _, _) or
+    FlowSummaryImpl::Private::Steps::summaryThroughStepTaint(e, _, _) or
+    FlowSummaryImpl::Private::Steps::summaryGetterStep(e, _, _, _) or
+    FlowSummaryImpl::Private::Steps::summarySetterStep(e, _, _, _)
+  }
+}
+
+/**
+ * A negative characteristic that filters out qualifiers that are classes (i.e. static calls). These
+ *are unlikely to have any non-trivial flow going into them.
+ */
+private class ClassQualifierCharacteristic extends CharacteristicsImpl::NotASinkCharacteristic {
+  ClassQualifierCharacteristic() { this = "class qualifier" }
+
+  override predicate appliesToEndpoint(Endpoint e) {
+    exists(Call c |
+      e.asExpr() = c.getQualifier() and
+      c.getQualifier() instanceof TypeAccess
+    )
+  }
+}
+
+/**
+ * A characteristic that limits candidates to parameters of methods that are recognized as `ModelApi`, iow., APIs that
+ * are considered worth modeling.
+ */
+private class ArgumentToLocalCall extends CharacteristicsImpl::UninterestingToModelCharacteristic {
+  ArgumentToLocalCall() { this = "argument to local call" }
+
+  override predicate appliesToEndpoint(Endpoint e) {
+    ApplicationCandidatesImpl::getCallable(e).fromSource()
+  }
+}
+
+/**
+ * A characteristic that avoids modeling endpoint that are passed to frameworks
+ * in application mode.
+ *
+ * It's much more economical to use framework mode for those.
+ */
+private class SkipFrameworkModeling extends CharacteristicsImpl::UninterestingToModelCharacteristic {
```

Comment:
I am somewhat concerned about this restriction, and am interested in hearing what challenges this caused when you were trying this out.

Reason: I think it is quite plausible that our field team would use application mode to understand the broad spectrum of possible unmodelled candidates within a codebase, and then use a combination of application and framework mode to model those candidates. If we have hardcoded restrictions on what endpoints get picked up by application mode, I can see this being very surprising for them to use, and difficult for us to debug why candidates are missing if we don't remember to look at this code.

Reply:
We added this restriction make sure we're not spending thousands of dollars of ML model cost for using application mode to model frameworks. We fully expect this implementation to not be perfect (a hard-coded set literal almost never is) — I think that application mode in the long run will need an input argument to define which packages to focus on (eg., by passing a wildcard) but that'd be a larger product/UX discussion.

That being said: if you want us to, we could remove this characteristic, but looking at the selected candidates makes me think we're better off filtering these frameworks out.

Reply:
Let's merge this as-is but I can work with you to make it configurable. We can use data extensions to specify whether to include/exclude certain "large framework" package patterns in each run of the extraction queries.

Reply:
The [recent triage](https://github.com/github/CodeML/issues/148#issuecomment-1579316995) of the first run of the new `application` mode on the top mrva repos yielded underwhelming results and if I understand this exclusion list correctly it's likely to be reason for it.

I share @adityasharad 's concerns about this - and in fact 

> I can see this being very surprising for them to use, and difficult for us to debug why candidates are missing if we don't remember to look at this code.

seems to be exactly what happened in the triage request.

I think our main focus for the moment should be to confirm the previous encouraging findings about the impact that `automodel` can have. Running application` mode at scale (without any restrictions) should help us there by flushing out existing gaps in our coverage of popular frameworks that are hopefully high impact (i.e. will lead to additional TP alerts). We could indeed also run `framework` mode on these frameworks to check/compare results - there are many potential things we could do here! But in the meantime, as mentioned in the stand-up it's probably safer to take such ideas out of this PR and discuss/evaluate them separately, along with the field team to assess what the right way forward is, once we are at that stage.

Reply:
We've removed the class for now, we can easily reproduce it later on (should the need arise). Additionally, @tausbn and I both think that @adityasharad's suggestion (to use data extensions for such a mechanism) is a promising idea.

[be6b1d8](https://github.com/github/codeql/pull/13239/commits/be6b1d8aaf1888e31612b7a8c92fe8f43d48982d)